cpp/src/text/minhash.cu

```diff
@@ -0,0 +1,187 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <nvtext/minhash.hpp>
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/copy.hpp>
+#include <cudf/detail/hashing.hpp>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/sequence.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/detail/utilities/device_atomics.cuh>
+#include <cudf/detail/utilities/hash_functions.cuh>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/execution_policy.h>
+#include <thrust/fill.h>
+#include <thrust/for_each.h>
+#include <thrust/iterator/counting_iterator.h>
+
+#include <limits>
+
+namespace nvtext {
+namespace detail {
+namespace {
+
+/**
+ * @brief Compute the minhash of each string for each seed
+ *
+ * This is a warp-per-string algorithm where parallel threads within a warp
+ * work on substrings of a single string row.
+ */
+struct minhash_fn {
+  cudf::column_device_view d_strings;
+  cudf::device_span<cudf::hash_value_type const> seeds;
+  cudf::size_type width;
+  cudf::hash_value_type* d_hashes;
+
+  __device__ void operator()(cudf::size_type idx)
+  {
+    auto const str_idx  = idx / cudf::detail::warp_size;
+    auto const lane_idx = idx % cudf::detail::warp_size;
+
+    if (d_strings.is_null(str_idx)) { return; }
+
+    auto const d_str    = d_strings.element<cudf::string_view>(str_idx);
+    auto const d_output = d_hashes + (str_idx * seeds.size());
+
+    // initialize hashes output for this string
+    if (lane_idx == 0) {
+      auto const init = d_str.empty() ? 0 : std::numeric_limits<cudf::hash_value_type>::max();
+      thrust::fill(thrust::seq, d_output, d_output + seeds.size(), init);
+    }
+    __syncwarp();
+
+    auto const begin = d_str.begin() + lane_idx;
+    auto const end   = [d_str, width = width] {
+      auto const length = d_str.length();
+      if (length > width) { return (d_str.end() - (width - 1)); }
+      return d_str.begin() + static_cast<cudf::size_type>(length > 0);
+    }();
+
+    // each lane hashes substrings of parts of the string
+    for (auto itr = begin; itr < end; itr += cudf::detail::warp_size) {
+      auto const offset = itr.byte_offset();
+      auto const hash_str =
+        cudf::string_view(d_str.data() + offset, (itr + width).byte_offset() - offset);
+
+      // hashing each seed on the same section of string is 10x faster than
+      // re-substringing (my new word) for each seed
```

Comment:
😉 I think this is clear enough. You could alternatively write "Using substrings in the outer loop and hash seeds in the inner loop is 10x faster than the other way around."
```suggestion
      // computing the substrings for each seed
```

Reply:
Some of these comments are just for you Bradley.

Reply:
I appreciate it! It made me smile. 😊